blogposts/2021/how-to-not-break-a-search-engine-or-what-i-learned-about-unglamorous-engineering.md

```diff
@@ -0,0 +1,297 @@
+challenging. This code will need checking and checking again. Not exactly
+glamorous.
+
+The changes weren't just about the act of parsing either. The new parser would
+produce a different internal representation of a query. Trees would represent
+our queries now, no longer just a sequential list of terms. This new
+representation needed to keep working with what our backend internals expect.
+That, or bugs galore.
+
+I made up my mind that, as far as I could help it, swapping our parsers wasn't
+going to break Sourcegraph. Not on my account.
+
+## How not to break a search engine
+
+Over the course of development I progressively tested new changes in four
+distinct ways. I wasn't following a checklist of things to do. What mattered
+was that I could convince myself that things worked as expected. I wanted
+to sleep well at night. Testing along these four different dimensions gave me
+confidence that I could expect few (if any) surprises. In general, the sort of
+testing for application rewrites or migration is going to depend on the
+context. I feel like the steps I took worked well, and they probably generalize
+well for migrating other systems that accept user input.
+
+### Part 1: Unit testing
+
+Unit testing was the obvious thing to do™ and familiar territory for a lot of
+developers, so I won't belabor this part. I thought of this as my frontline
+defense for testing correctness. I reused some of our existing parser tests
+(which served as a rough specification of things to get right) and also added a
+lot of additional tests for new parts of the syntax. You bet there's test
+coverage.
+
+<img src="https://storage.googleapis.com/sourcegraph-assets/about.sourcegraph.com/blog/2021/parser-coverage.png" style="width: 300px" alt="Parser file unit test coverage">
+
+### Part 2: Integration testing
+
+Once I wrote a feature-flagged code path to use the new parser, it was time to
+test whether Sourcegraph behaved functionally similar. Integration tests ran a
+search query through a running instance of the Sourcegraph application and
+checked whether search results were expected. This testing phase was so
+important because the parser produced a new and different internal
+representation. I'd abstracted out a common interface for our backend to access
+the new data structure under the feature-flagged code path to test on.
+
+<figure style="width: 800px; margin: auto;">
+  <img src="https://storage.googleapis.com/sourcegraph-assets/about.sourcegraph.com/blog/2021/integration.png" alt="Integration testing diagram">
+  <a id="fig-1"><figcaption>
+  <i>Fig. 1: Integration testing abstracts a common interface for our backend
+  to access query values.
+  We test that the backend produces the same search results for
+  simple queries (ones that don't have, e.g., </i>
+  <code>or</code>
+  <i>-operators), irrespective of whether those values originate from the existing parser's output or the new one.</i></figcaption></a>
+</figure>
```

Comment:
@rvantonder This isn't a blocking question, but I don't think we need to call this "Fig 1" as there are no further images labelled "Fig". 

Also, the drop shadow is great for screengrabs, but on diagrams with a white background it looks jarring to me to have the shadow break up the background of the image + the rest of the post unnecessarily. I can't see what is creating the drop shadow myself to remove it. If it's a pain to change we can just leave as is!

Reply:
The Fig label is left over some something I tried and reverted. We should remove :-)

The backdrop is I think a default gatsby CSS thing. I think we'd need to add a custom style attr to get rid of the backdrop. Agree it would look better without it, but would need some hacking.